packages/firestore/src/local/local_store.ts

```diff
@@ -89,6 +95,40 @@ export interface QueryResult {
   readonly remoteKeys: DocumentKeySet;
 }
 
+/**
+ * Base of local store implementations. This is mainly used to define fields
+ * that are used by tree-shakeable free functions that implement optional
+ * local store features.
+ */
+class LocalStoreBase {
```

Comment:
So the idea behind the pattern that is used in Datastore/DatastoreImpl (and just as of today, also in Target/TargetImpl) is to have a main class (in this case LocalStore) that doesn't change its public API.  

So you would keep LocalStore as is (and continue to export it), and then you would create LocalStoreImpl that makes fields public that you only need to be public in this module. Then you would add a `newLocalStore` function that creates a LocalStoreImpl, but is typed to return a LocalStore. You then end up with a module that has a bunch of public members that are not leaked externally (in our code base).

Reply:
Sorry I misunderstood earlier.

As discussed offline, we will need to make LocalStore and SyncEngine interfaces to achieve this because of existing MultiTab* implementations, which is best done from master.